File: banquet.go
package banquet

import (
	"fmt"
	"log"
	"net/url"
	"strconv"
	"strings"
)

var verbose bool

// SetVerbose enables or disables verbose logging for the package.
func SetVerbose(v bool) {
	verbose = v
}

// IsVerbose returns true if verbose logging is enabled.
func IsVerbose() bool {
	return verbose
}

// Banquet package and Banquet struct are designed to provide a URL standardization and parsing framework for tabular data.
// What will banquet have the url.URL doesn't?
// 1. Where clause
// 2. Select clause
// 3. Sort clause
// 4. Limit clause
// 5. Offset clause
// 6. Group by clause
// 7. Having clause
// 8. Order by clause
// 9. Slice Notation [2:30] inspired by python and golang to signal subset of rows.
// 10. Special characters to signal sorting direction, ascending or descending.
// 11. Repurpose userinfo to signal authentication.
// 12. Repurpose path to signal file path or db path or object key or zip file path.
// 13. Repurpose query to signal query parameters.
// 14. Lots of tolerance for unescaped characters.
// ColumnPath is the path to the column.
type Banquet struct {
	*url.URL
	Where         string
	Table         string   // table name to go in FROM clause parsed from request url
	Select        []string // columns to select.  empty or * means all columns
	SortDirection string   // refactor this to mean ASC or DESC. We have OrderBy for previous Sort meaning.
	Limit         string
	Offset        string
	GroupBy       string
	Having        string
	OrderBy       string
	DataSetPath   string // Server needs this to respond with the downloadable, convedata set. Excel, CSV, eventually BigQuery dataset.

	ColumnPath string // Formatted table/column1, column2. Empty means select * from dataset that only has one table..
	// fields below are for internal use
	rawurl string
	path   string
}

const (
	// Sort direction tokens
	ASC  = "+" // token to signal the following column is sorted ascending
	DESC = "-" // token to signal the following column is sorted descending

)

/*
// General form of a URL:
//
//	[scheme:][//[userinfo@]host][/]path[?query][#fragment]
*/
/* General form of a Banquet:
//
Familiar form:
//	[scheme:][//[userinfo@]host][/]path/to/dataset/table/column1,column2,column3...?[where][select][sort][limit][offset][groupby][having][orderby][slice][#fragment]
Since I can't find a way to signal a priori that path part is table vs column name vs file path, we will use a convention
Canonical form:
[scheme:][//[userinfo@]host][/]path/to/dataset;table/column1,column2,column3...?[where][select][sort][limit][offset][groupby][having][orderby][slice][#fragment]
Canonical form if for single table sources:
[scheme:][//[userinfo@]host][/]path/to/dataset;column1,column2,column3...?[where][select][sort][limit][offset][groupby][having][orderby][slice][#fragment]

Ambiguous form:
[scheme:][//[userinfo@]host][/]path/to/dataset/tableorcolumn?[where][select][sort][limit][offset][groupby][having][orderby][slice][#fragment]

// List of prefixes to support
//  + and - prefixed to column name signals sorting on that column.
//  [number:number] signals slice notation. It gets translated to limit and offset.
//  TBD signals group by.
//  TBD signals having.
//  TBD signals order by.
// List of suffixes to support
//  Suffixes such as file ending are so important to parsing that I will probably avoid more suffixes.
*/

/*Example Banquet Unparseable	reqURL := "https://localhost:8080/gs:/matrix@bucket.appspot.com:8080/some/file/path.csv/column1,column2/^column3?orderid=1"
Files found in tests:
reqURL := "/gs://wc2022-356423.appspot.com/Expenses.csv/%5EDescription"
reqURL := "gs:///wc2022-356423.appspot.com/Expenses.csv/^Description"
testURL := "https://raw.githubusercontent.com/uiuc-cse/data-fa14/gh-pages/data/iris.csv"
testURL := "https://raw.githubusercontent.com/holtzy/data_to_viz/master/Example_dataset/1_OneNum.csv"
testURL := "https://raw.githubusercontent.com/uiuc-cse/data-fa14/gh-pages/data/iris.csv"
reqURL := "https://localhost:8080/gs:/matrix@bucket.appspot.com:8080/some/file/path.csv/column1,column2/^column3?orderid=1"
reqURL := "https://localhost:8080/gs:/some/file/path.csv/column1,column2/^column3?where=1=1"
reqURL := "gs:/maverick@buckt1.appspot.com/some/file/path.csv/column1,column2/^column3?where=1=1"
reqURL := "http://localhost:8080/sample%20data/demo%20mavgo%20flight/Expenses.csv"
reqURL := "http://localhost:8080/data.csv/^column1"
*/

// A big UI incompatibility is that I hate url escape sequences.

// So unescape tolerant parsing need to be here. ! PathUnescape already exists. !
// Trims leading slash, fixes scheme format, but expecting more cleanings as we go.
func CleanUrl(rawurl string) string {
	// housekeeping before url.Parse
	if rawurl == "/" {
		return "."
	}
	rawurl = strings.TrimPrefix(rawurl, "/")

	// Ensure standard scheme format (e.g., gs:/ -> gs://) for proper authority parsing
	if idx := strings.Index(rawurl, ":/"); idx != -1 {
		if !strings.HasPrefix(rawurl[idx:], "://") {
			rawurl = strings.Replace(rawurl, ":/", "://", 1)
		}
	}

	return rawurl
}

// go DOES NOT support override of URL.Parse so instead we will use a factory function.
func ParseBanquet(rawurl string) (*Banquet, error) {
	if verbose {
		log.Printf("[BANQUET] Parsing URL: %s", rawurl)
	}
	// Standardize/Clean the URL (trim leading slash, fix scheme)
	rawurl = CleanUrl(rawurl)

	if verbose {
		log.Printf("[BANQUET] Cleaned URL: %s", rawurl)
	}

	u, err := url.Parse(rawurl)
	if err != nil {
		if verbose {
			log.Printf("[BANQUET] URL parse error: %v", err)
		}
		return nil, err
	}

	b := &Banquet{
		URL:    u,
		rawurl: rawurl,
	}

	b.DataSetPath, b.Table, b.ColumnPath = parseDataSetColumnPath(b.Path)
	if verbose {
		log.Printf("[BANQUET] DataSetPath: %s, Table: %q, ColumnPath: %s", b.DataSetPath, b.Table, b.ColumnPath)
	}

	// Table parsing logic - fallback to heuristic only if not explicitly set via semicolon
	if b.Table == "" {
		b.Table = parseTable(b.ColumnPath)
		if verbose {
			log.Printf("[BANQUET] Table identified via heuristic: %s", b.Table)
		}
	}

	// Populate fields using private parsers
	b.Select = ParseSelect(b.ColumnPath)
	if verbose {
		log.Printf("[BANQUET] Selected columns: %v", b.Select)
	}

	// Fix: if Select is just the Table name, change to *
	if len(b.Select) == 1 && b.Select[0] == b.Table {
		b.Select = []string{"*"}
	}

	// Combine query params 'where' and path conditions
	queryWhere := parseWhere(b.RawQuery)
	pathWhere := parsePathConditions(b.ColumnPath)

	if pathWhere != "" {
		if queryWhere != "" {
			b.Where = queryWhere + " AND " + pathWhere
		} else {
			b.Where = pathWhere
		}
	} else {
		b.Where = queryWhere
	}

	if verbose && b.Where != "" {
		log.Printf("[BANQUET] effective WHERE: %s", b.Where)
	}

	b.GroupBy = ParseGroupBy(b.Path, b.RawQuery)

	b.Limit = parseLimit(b.RawQuery, b.Path)
	b.Offset = parseOffset(b.RawQuery, b.Path)
	b.Having = parseHaving(b.RawQuery)
	if ob, dir := parseOrderBy(b.ColumnPath, b.RawQuery); ob != "" {
		b.OrderBy = ob
		if dir != "" {
			b.SortDirection = dir
		}
	}

	return b, nil
}

func FmtPrintln(b *Banquet) {
	fmt.Printf(`rawurl: %s
Scheme: %s
  Host:   %s
  DataSetPath: %s
  ColumnPath: %s
  RawQuery: %s
  Table:      %q
`, b.rawurl, b.Scheme, b.Host, b.DataSetPath, b.ColumnPath, b.RawQuery, b.Table)
}

func FmtSprintf(b *Banquet) string {
	return fmt.Sprintf(`rawurl: %s\n
  S: %s H: %s DP: %sCP: %sRQ:%sTB:%q
`, b.rawurl, b.Scheme, b.Host, b.DataSetPath, b.ColumnPath, b.RawQuery, b.Table)
}

// Don't create func Parse(rawurl string) (*Banquet, error) we basically have that in url.Parse

// ParseNested is a factory function that creates a new Banquet from Banquet URL nested in an outer URL.
// Expecting this a lot from http.Request.
func ParseNested(rawURL string) (*Banquet, error) {
	// 1. Parse the outer envelope
	// If rawURL is http://localhost..., url.Parse works.
	// If rawURL is just /http..., we need to trim prefix, but not if it's just "/"
	if rawURL != "/" {
		rawURL = strings.TrimPrefix(rawURL, "/")
	}

	outer, err := url.Parse(rawURL)
	if err != nil {
		// If outer parse fails, we might just try to treat the whole thing as an inner url?
		// But usually this means it's really malformed.
		return nil, err
	}

	// Key Fix: Use EscapedPath() to get the path segment exactly as it was on the wire (checking for %25 etc)
	// If we use outer.Path, %25 is decoded to %, which might form invalid escapes later.
	inner := outer.EscapedPath()

	// If strings.HasPrefix(inner, "/") -> trimming?
	// ParseBanquet trims leading slash.

	if outer.RawQuery != "" {
		inner += "?" + outer.RawQuery
	}

	b, err := ParseBanquet(inner)
	if err != nil {
		fmt.Printf("Error parsing inner URL '%s': %v. Continuing with raw URL.\n", inner, err)
		// Return partial Banquet with just the raw extraction
		return &Banquet{
			URL:    &url.URL{Path: inner}, // Best effort
			rawurl: inner,
		}, nil
	}
	return b, nil
}

// Internal parsing functions
func parseDataSetColumnPath(rawpath string) (datasetPath string, table string, columnPath string) {
	// If rawpath contains semicolons, we use explicit tier parsing: dataset;table;columns
	if strings.Contains(rawpath, ";") {
		parts := strings.SplitN(rawpath, ";", 3)
		datasetPath = parts[0]
		if len(parts) > 1 {
			table = parts[1]
		}
		if len(parts) > 2 {
			columnPath = parts[2]
		}
		return
	}

	// if there is no ";" then use existing file extension logic to split path into dataset path and column path
	parts := strings.Split(rawpath, "/")
	for i, part := range parts {
		if strings.HasSuffix(part, ".zip") ||
			strings.HasSuffix(part, ".csv") ||
			strings.HasSuffix(part, ".sqlite") ||
			strings.HasSuffix(part, ".db") ||
			strings.HasSuffix(part, ".xlsx") ||
			strings.HasSuffix(part, ".json") ||
			(strings.HasSuffix(part, ".html") && part != "test.html") ||
			strings.HasSuffix(part, ".txt") {

			datasetPath = strings.Join(parts[:i+1], "/")
			if i+1 < len(parts) {
				columnPath = strings.Join(parts[i+1:], "/")
			}
			return datasetPath, "", columnPath
		}
	}
	return rawpath, "", ""
}

// getSegments identifies the part of the path that contains columns or conditions
func getSegments(columnPath string) []string {
	parts := strings.Split(columnPath, "/")
	if len(parts) == 0 || (len(parts) == 1 && parts[0] == "") {
		return nil
	}

	// Find the first segment that contains clear column/sort/condition indicators
	firstClearSegment := -1
	for i, part := range parts {
		if strings.Contains(part, ",") ||
			strings.HasPrefix(part, ASC) ||
			strings.HasPrefix(part, DESC) ||
			strings.Contains(part, "!=") ||
			(strings.HasPrefix(part, "[") && strings.Contains(part, ":")) {
			firstClearSegment = i
			break
		}
	}

	if firstClearSegment != -1 {
		return parts[firstClearSegment:]
	}

	// If no indicators, just return the last segment (could be table or single column)
	return parts[len(parts)-1:]
}

func ParseSelect(columnPath string) []string {
	segments := getSegments(columnPath)
	if len(segments) == 0 {
		return []string{"*"}
	}

	var collected []string
	for _, segment := range segments {
		// Ignore slice notation
		if strings.HasPrefix(segment, "[") && strings.HasSuffix(segment, "]") && strings.Contains(segment, ":") {
			continue
		}
		if segment == "" {
			continue
		}
		cols := strings.Split(segment, ",")
		for _, col := range cols {
			// Ignore conditions
			if strings.Contains(col, "!=") {
				continue
			}

			// If it has a sort prefix, it's for ordering, not necessarily for selection.
			// In banquet, table/+id implies SELECT * FROM table ORDER BY id ASC.
			if strings.HasPrefix(col, ASC) || strings.HasPrefix(col, DESC) {
				continue
			}

			// Clean up slice notation
			if idx := strings.Index(col, "["); idx != -1 {
				col = col[:idx]
			}

			// Basic cleanup
			col = strings.TrimSpace(col)
			if col != "" {
				collected = append(collected, col)
			}
		}
	}

	if len(collected) == 0 {
		return []string{"*"}
	}

	return collected
}

func parsePathConditions(columnPath string) string {
	segments := getSegments(columnPath)
	if len(segments) == 0 {
		return ""
	}

	var conditions []string
	for _, segment := range segments {
		if segment == "" {
			continue
		}
		// Conditions can be comma separated? e.g. col1!=val1,col2!=val2
		// Assuming yes since ParseSelect splits by comma.
		parts := strings.Split(segment, ",")
		for _, part := range parts {
			if strings.Contains(part, "!=") {
				// Split
				kv := strings.SplitN(part, "!=", 2)
				if len(kv) == 2 {
					col := strings.TrimSpace(kv[0])
					val := strings.TrimSpace(kv[1])

					// URL Decode value
					decodedVal, err := url.QueryUnescape(val)
					if err == nil {
						val = decodedVal
					}

					// Quote if not number
					if _, err := strconv.ParseFloat(val, 64); err != nil {
						// Quote single quotes for SQL safety
						val = strings.ReplaceAll(val, "'", "''")
						val = "'" + val + "'"
					}

					conditions = append(conditions, fmt.Sprintf("%s != %s", col, val))
				}
			}
		}
	}

	if len(conditions) == 0 {
		return ""
	}
	return strings.Join(conditions, " AND ")
}

func parseWhere(query string) string {
	if query == "" {
		return ""
	}
	// Simple extraction of 'where' parameter
	// url.ParseQuery is too strict for Banquet's "unescape tolerant" goal.
	params := strings.Split(query, "&")
	for _, p := range params {
		if strings.HasPrefix(p, "where=") {
			val := strings.TrimPrefix(p, "where=")
			// Try to unescape but if it fails, just return the raw value
			if decoded, err := url.QueryUnescape(val); err == nil {
				return decoded
			}
			return val
		}
	}
	return ""
}

func ParseGroupBy(path string, query string) string {
	// check query first
	v, _ := url.ParseQuery(query)
	if g := v.Get("groupby"); g != "" {
		return g
	}

	// check path for (expression)
	if strings.Contains(path, "(") && strings.Contains(path, ")") {
		start := strings.Index(path, "(")
		end := strings.Index(path, ")")
		if start < end {
			return path[start+1 : end]
		}
	}
	return ""
}

// parseSort function deleted as it is superseded by parseOrderBy and parseSortStr
func parseTable(columnPath string) string {
	if columnPath == "" {
		return ""
	}
	// Trim slashes and split
	parts := strings.Split(strings.Trim(columnPath, "/"), "/")
	if len(parts) == 0 || parts[0] == "" {
		return ""
	}

	first := parts[0]

	// Indicators that this segment is a selector (column list, sort, or filter) and NOT a table name.
	// Comparison operators, commas, and sort prefixes are clear structural clues.
	if strings.Contains(first, ",") ||
		strings.HasPrefix(first, ASC) ||
		strings.HasPrefix(first, DESC) ||
		strings.Contains(first, "!=") ||
		strings.Contains(first, "=") ||
		strings.Contains(first, ">") ||
		strings.Contains(first, "<") ||
		(strings.HasPrefix(first, "[") && strings.Contains(first, ":")) {
		return ""
	}

	// Heuristic: If it's not a selector, it's likely a table or resource name.
	// This works for SQLite (db.sqlite/users) and handles CSV (file.csv/col1,col2) correctly.
	// For ambiguous single segments like /column, it will tentatively return it as a table;
	// downstream logic in ParseBanquet handles the table/column overlap.
	return first
}

func parseLimit(query string, path string) string {
	v, _ := url.ParseQuery(query)
	if l := v.Get("limit"); l != "" {
		return l
	}
	// Check path for slice notation [offset:limit]
	if limit, _ := parseSlice(path); limit != "" {
		return limit
	}
	return ""
}

func parseOffset(query string, path string) string {
	v, _ := url.ParseQuery(query)
	if o := v.Get("offset"); o != "" {
		return o
	}
	_, offset := parseSlice(path)
	return offset
}

func parseHaving(query string) string {
	v, _ := url.ParseQuery(query)
	return v.Get("having")
}

func parseOrderBy(columnPath string, query string) (string, string) {
	v, _ := url.ParseQuery(query)
	if ob := v.Get("orderby"); ob != "" {
		return ob, ""
	}

	// check path parts
	parts := strings.Split(columnPath, "/")
	for _, part := range parts {
		cols := strings.Split(part, ",")
		for _, col := range cols {
			col = strings.TrimSpace(col)
			// Strip slice notation
			if idx := strings.Index(col, "["); idx != -1 {
				col = col[:idx]
			}
			if strings.HasPrefix(col, ASC) {
				return strings.TrimPrefix(col, ASC), "ASC"
			}
			if strings.HasPrefix(col, DESC) {
				return strings.TrimPrefix(col, DESC), "DESC"
			}
		}
	}
	return "", ""
}

func parseSlice(pathStr string) (string, string) {
	if !strings.HasSuffix(pathStr, "]") {
		return "", ""
	}
	idx := strings.LastIndex(pathStr, "[")
	if idx == -1 {
		return "", ""
	}
	content := pathStr[idx+1 : len(pathStr)-1]
	parts := strings.Split(content, ":")
	if len(parts) != 2 {
		return "", ""
	}

	startStr := strings.TrimSpace(parts[0])
	endStr := strings.TrimSpace(parts[1])

	start := 0
	end := 0
	hasLimit := false

	if startStr != "" {
		s, err := strconv.Atoi(startStr)
		if err != nil {
			return "", ""
		}
		start = s
	}

	if endStr != "" {
		e, err := strconv.Atoi(endStr)
		if err != nil {
			return "", ""
		}
		end = e
		hasLimit = true
	}

	offset := start
	limit := ""

	if hasLimit {
		l := end - start
		if l < 0 {
			l = 0
		}
		limit = strconv.Itoa(l)
	}

	return limit, strconv.Itoa(offset)
}

File: dart/lib/banquet.dart
/// Package banquet provides a framework for standardizing and parsing URLs for
/// tabular data access. It supports querying datasets (CSV, SQLite, etc.) via
/// URL paths and query parameters, allowing typical SQL-like operations such as
/// selection, filtering, sorting, limiting, and grouping directly from the URL.
///
/// Banquet Classifications & URL Structure:
///
/// 1. Flat (1-tier):          path/to/dataset;;Column
/// 2. Nested Table (2-tier):  path/to/dataset;Table
/// 3. Nested Column (3-tier): path/to/dataset;Table;Column
///
/// Fallback Convention (semicolon-less):
/// If no semicolons are present, the package heuristically determines the table
/// or column based on file extensions.
///
/// Supported Prefixes/Suffixes:
/// - Sort: +column (ASC), -column (DESC)
/// - Slice: [start:end] (translated to LIMIT/OFFSET)
library banquet;

bool _verbose = false;

/// Sets verbose logging for the package.
void setVerbose(bool v) => _verbose = v;

/// Returns true if verbose logging is enabled.
bool isVerbose() => _verbose;

/// ASC is the prefix token to signal ascending sort order.
const String kAsc = '+';

/// DESC is the prefix token to signal descending sort order.
const String kDesc = '-';

// ---------------------------------------------------------------------------
// Banquet
// ---------------------------------------------------------------------------

/// Banquet represents a parsed URL request for tabular data.
/// It wraps [Uri] with SQL-like clauses derived from the path and query params.
class Banquet {
  /// The underlying parsed URI (replaces Go's *url.URL).
  final Uri uri;

  String where = '';
  String table = '';
  List<String> select = const ['*'];
  String sortDirection = '';
  String limit = '';
  String offset = '';
  String groupBy = '';
  String having = '';
  String orderBy = '';

  /// Path to the source dataset file (e.g., .csv, .sqlite).
  String dataSetPath = '';

  /// The remaining path segment after the dataset, containing columns, sort
  /// instructions, or conditions.
  String columnPath = '';

  // Internal fields
  final String _rawurl;

  Banquet._({
    required this.uri,
    required String rawurl,
    this.dataSetPath = '',
    this.table = '',
    this.columnPath = '',
  }) : _rawurl = rawurl;

  // Convenience passthrough getters to mirror Go's embedded *url.URL fields.
  String get scheme => uri.scheme;
  /// Returns the host exactly as Go's url.URL.Host does – including port when present.
  /// e.g. 'bucket.appspot.com:8080' not just 'bucket.appspot.com'.
  String get host => uri.hasPort ? '${uri.host}:${uri.port}' : uri.host;
  String get path => uri.path;
  String get rawQuery => uri.query;
  String port() => uri.hasPort ? '${uri.port}' : '';


  /// Returns the userinfo username (mirrors Go's b.User.Username()).
  String username() => uri.userInfo.contains(':')
      ? uri.userInfo.split(':').first
      : uri.userInfo;

  @override
  String toString() =>
      'rawurl: $_rawurl\n'
      '  S: $scheme H: $host DP: $dataSetPath CP: $columnPath RQ: $rawQuery TB: "$table"';
}

// ---------------------------------------------------------------------------
// CleanUrl
// ---------------------------------------------------------------------------

/// CleanUrl prepares a raw URL string for standard parsing.
/// Mirrors [banquet.CleanUrl] in Go exactly.
String cleanUrl(String rawurl) {
  if (rawurl == '/') return '.';
  rawurl = rawurl.replaceFirst(RegExp(r'^/'), '');

  final schemeIdx = rawurl.indexOf(':/');
  if (schemeIdx != -1) {
    // Ensure :/ becomes :// (e.g. gs:/ -> gs://)
    if (!rawurl.substring(schemeIdx).startsWith('://')) {
      rawurl = rawurl.replaceFirst(':/', '://');
    }
  } else {
    // No scheme separator present – guard against colon-in-first-segment issue.
    if (!rawurl.contains('://') && rawurl.contains(':')) {
      final slashIdx = rawurl.indexOf('/');
      final colonIdx = rawurl.indexOf(':');
      if (slashIdx == -1 || colonIdx < slashIdx) {
        return './$rawurl';
      }
    }
  }
  return rawurl;
}

// ---------------------------------------------------------------------------
// ParseBanquet
// ---------------------------------------------------------------------------

/// Parses a raw URL string into a [Banquet] object.
/// Mirrors [banquet.ParseBanquet] in Go exactly.
Banquet parseBanquet(String rawurl) {
  if (_verbose) print('[BANQUET] Parsing URL: $rawurl');

  rawurl = cleanUrl(rawurl);
  if (_verbose) print('[BANQUET] Cleaned URL: $rawurl');

  final Uri u;
  try {
    u = Uri.parse(rawurl);
  } catch (e) {
    if (_verbose) print('[BANQUET] URL parse error: $e');
    rethrow;
  }

  final b = Banquet._(uri: u, rawurl: rawurl);

  final (dsPath, tbl, colPath) = _parseDataSetColumnPath(u.path);
  b.dataSetPath = dsPath;
  b.table = tbl;
  b.columnPath = colPath;

  if (_verbose) {
    print('[BANQUET] DataSetPath: ${b.dataSetPath}, Table: "${b.table}", ColumnPath: ${b.columnPath}');
  }

  // Fallback heuristic table identification
  if (b.table.isEmpty) {
    b.table = _parseTable(b.columnPath);
    if (_verbose) print('[BANQUET] Table identified via heuristic: ${b.table}');
  }

  // Strip slice notation from table
  final bracketIdx = b.table.indexOf('[');
  if (bracketIdx != -1) {
    b.table = b.table.substring(0, bracketIdx);
  }

  b.select = parseSelect(b.columnPath);
  if (_verbose) print('[BANQUET] Selected columns: ${b.select}');

  // If Select is just the table name, default to *
  if (b.select.length == 1 && b.select[0] == b.table) {
    b.select = ['*'];
  }

  final queryWhere = _parseWhere(u.query);
  final pathWhere = _parsePathConditions(b.columnPath);

  if (pathWhere.isNotEmpty) {
    b.where =
        queryWhere.isNotEmpty ? '$queryWhere AND $pathWhere' : pathWhere;
  } else {
    b.where = queryWhere;
  }

  if (_verbose && b.where.isNotEmpty) {
    print('[BANQUET] effective WHERE: ${b.where}');
  }

  b.groupBy = parseGroupBy(u.path, u.query);
  b.limit = _parseLimit(u.query, u.path);
  b.offset = _parseOffset(u.query, u.path);
  b.having = _parseHaving(u.query);

  final (ob, dir) = _parseOrderBy(b.columnPath, u.query);
  if (ob.isNotEmpty) {
    b.orderBy = ob;
    if (dir.isNotEmpty) b.sortDirection = dir;
  }

  return b;
}

// ---------------------------------------------------------------------------
// ParseNested
// ---------------------------------------------------------------------------

/// Extracts and parses a Banquet URL that wraps an inner URL.
/// Mirrors [banquet.ParseNested] in Go exactly.
Banquet parseNested(String rawURL) {
  if (rawURL != '/') {
    rawURL = rawURL.replaceFirst(RegExp(r'^/'), '');
  }

  final Uri outer;
  try {
    outer = Uri.parse(rawURL);
  } catch (e) {
    rethrow;
  }

  // Use the raw (encoded) path, equivalent to outer.EscapedPath() in Go.
  String inner = outer.path;

  if (outer.query.isNotEmpty) {
    inner = '$inner?${outer.query}';
  }

  try {
    return parseBanquet(inner);
  } catch (e) {
    print("Error parsing inner URL '$inner': $e. Continuing with raw URL.");
    // Best-effort fallback (mirrors Go's partial Banquet return).
    return Banquet._(uri: Uri(path: inner), rawurl: inner);
  }
}

// ---------------------------------------------------------------------------
// ParseSelect (exported)
// ---------------------------------------------------------------------------

/// Parses the column path to determine which columns to select.
/// Mirrors [banquet.ParseSelect] in Go exactly.
List<String> parseSelect(String columnPath) {
  final segments = _getSegments(columnPath);
  if (segments.isEmpty) return ['*'];

  final collected = <String>[];
  for (final segment in segments) {
    if (segment.isEmpty) continue;

    final cols = segment.split(',');
    for (var col in cols) {
      // Strip slice notation only if it contains ':'
      final idx = col.indexOf('[');
      if (idx != -1 && col.substring(idx).contains(':')) {
        col = col.substring(0, idx);
      }

      col = col.trim();
      if (col.isEmpty) continue;

      // Conditions are not selections
      if (col.contains('!=')) continue;

      // Sort prefixes mean this col is for ordering, not selection
      if (col.startsWith(kAsc) || col.startsWith(kDesc)) continue;

      collected.add(col);
    }
  }

  return collected.isEmpty ? ['*'] : collected;
}

// ---------------------------------------------------------------------------
// ParseGroupBy (exported)
// ---------------------------------------------------------------------------

/// Parses the group-by clause from the path or query.
/// Mirrors [banquet.ParseGroupBy] in Go exactly.
String parseGroupBy(String path, String query) {
  final params = Uri.splitQueryString(query);
  final g = params['groupby'] ?? '';
  if (g.isNotEmpty) return g;

  // Check path for (expression)
  if (path.contains('(') && path.contains(')')) {
    final start = path.indexOf('(');
    final end = path.indexOf(')');
    if (start < end) return path.substring(start + 1, end);
  }
  return '';
}

// ---------------------------------------------------------------------------
// FmtPrintln / FmtSprintf helpers
// ---------------------------------------------------------------------------

void fmtPrintln(Banquet b) {
  print('rawurl: ${b._rawurl}\n'
      'Scheme: ${b.scheme}\n'
      '  Host:   ${b.host}\n'
      '  DataSetPath: ${b.dataSetPath}\n'
      '  ColumnPath: ${b.columnPath}\n'
      '  RawQuery: ${b.rawQuery}\n'
      '  Table:      "${b.table}"');
}

String fmtSprintf(Banquet b) =>
    'rawurl: ${b._rawurl}\\n'
    '  S: ${b.scheme} H: ${b.host} DP: ${b.dataSetPath}'
    'CP: ${b.columnPath}RQ:${b.rawQuery}TB:"${b.table}"';

// ---------------------------------------------------------------------------
// Private helpers (mirrors unexported Go functions)
// ---------------------------------------------------------------------------

/// Mirrors [banquet.parseDataSetColumnPath].
(String datasetPath, String table, String columnPath) _parseDataSetColumnPath(
    String rawpath) {
  if (rawpath.contains(';')) {
    final parts = rawpath.split(';');
    final datasetPath = parts[0];
    final table = parts.length > 1 ? parts[1] : '';
    final columnPath = parts.length > 2 ? parts.sublist(2).join(';') : '';
    return (datasetPath, table, columnPath);
  }

  // Extension-based splitting (no semicolons)
  final parts = rawpath.split('/');
  const knownExts = ['.zip', '.csv', '.sqlite', '.db', '.xlsx', '.json', '.txt'];

  for (var i = 0; i < parts.length; i++) {
    final part = parts[i];
    final isHtml =
        part.endsWith('.html') && part != 'test.html';
    final hasExt = knownExts.any((e) => part.endsWith(e)) || isHtml;
    if (hasExt) {
      final datasetPath = parts.sublist(0, i + 1).join('/');
      final columnPath =
          i + 1 < parts.length ? parts.sublist(i + 1).join('/') : '';
      return (datasetPath, '', columnPath);
    }
  }
  return (rawpath, '', '');
}

/// Mirrors [banquet.getSegments].
List<String> _getSegments(String columnPath) {
  final parts = columnPath.split('/');
  if (parts.isEmpty || (parts.length == 1 && parts[0].isEmpty)) return [];

  var firstClear = -1;
  for (var i = 0; i < parts.length; i++) {
    final part = parts[i];
    if (part.contains(',') ||
        part.startsWith(kAsc) ||
        part.startsWith(kDesc) ||
        part.contains('!=') ||
        (part.startsWith('[') && part.contains(':'))) {
      firstClear = i;
      break;
    }
  }

  if (firstClear != -1) return parts.sublist(firstClear);
  return [parts.last];
}

/// Mirrors [banquet.parsePathConditions].
String _parsePathConditions(String columnPath) {
  final segments = _getSegments(columnPath);
  if (segments.isEmpty) return '';

  final conditions = <String>[];
  for (final segment in segments) {
    if (segment.isEmpty) continue;
    final parts = segment.split(',');
    for (final part in parts) {
      if (part.contains('!=')) {
        final kv = part.split('!=');
        if (kv.length == 2) {
          final col = kv[0].trim();
          var val = kv[1].trim();

          // URL-decode the value (tolerant)
          try {
            val = Uri.decodeQueryComponent(val);
          } catch (_) {}

          // Quote if not a number
          if (double.tryParse(val) == null) {
            val = val.replaceAll("'", "''");
            val = "'$val'";
          }
          conditions.add('$col != $val');
        }
      }
    }
  }
  return conditions.isEmpty ? '' : conditions.join(' AND ');
}

/// Mirrors [banquet.parseWhere].
String _parseWhere(String query) {
  if (query.isEmpty) return '';
  // Tolerant split – same approach as Go (manual split, not Uri.splitQueryString).
  final params = query.split('&');
  for (final p in params) {
    if (p.startsWith('where=')) {
      final val = p.substring('where='.length);
      try {
        return Uri.decodeQueryComponent(val);
      } catch (_) {
        return val;
      }
    }
  }
  return '';
}

/// Mirrors [banquet.parseTable].
String _parseTable(String columnPath) {
  if (columnPath.isEmpty) return '';

  final parts = columnPath
      .replaceAll(RegExp(r'^/+|/+$'), '')
      .split('/');
  if (parts.isEmpty || parts[0].isEmpty) return '';

  final first = parts[0];

  if (first.contains(',') ||
      first.startsWith(kAsc) ||
      first.startsWith(kDesc) ||
      first.contains('!=') ||
      first.contains('=') ||
      first.contains('>') ||
      first.contains('<') ||
      (first.startsWith('[') && first.contains(':'))) {
    return '';
  }
  return first;
}

/// Mirrors [banquet.parseLimit].
String _parseLimit(String query, String path) {
  final params = Uri.splitQueryString(query);
  final l = params['limit'] ?? '';
  if (l.isNotEmpty) return l;
  final (limit, _) = _parseSlice(path);
  return limit;
}

/// Mirrors [banquet.parseOffset].
String _parseOffset(String query, String path) {
  final params = Uri.splitQueryString(query);
  final o = params['offset'] ?? '';
  if (o.isNotEmpty) return o;
  final (_, offset) = _parseSlice(path);
  return offset;
}

/// Mirrors [banquet.parseHaving].
String _parseHaving(String query) {
  final params = Uri.splitQueryString(query);
  return params['having'] ?? '';
}

/// Mirrors [banquet.parseOrderBy].
(String orderBy, String direction) _parseOrderBy(
    String columnPath, String query) {
  final params = Uri.splitQueryString(query);
  final ob = params['orderby'] ?? '';
  if (ob.isNotEmpty) return (ob, '');

  final parts = columnPath.split('/');
  for (final part in parts) {
    final cols = part.split(',');
    for (var col in cols) {
      col = col.trim();
      final bracketIdx = col.indexOf('[');
      if (bracketIdx != -1) col = col.substring(0, bracketIdx);

      if (col.startsWith(kAsc)) {
        return (col.substring(kAsc.length), 'ASC');
      }
      if (col.startsWith(kDesc)) {
        return (col.substring(kDesc.length), 'DESC');
      }
    }
  }
  return ('', '');
}

/// Mirrors [banquet.parseSlice].
/// Returns (limit, offset) strings, matching Go's (string, string) return.
(String limit, String offset) _parseSlice(String pathStr) {
  final startIdx = pathStr.lastIndexOf('[');
  if (startIdx == -1) return ('', '');

  final sub = pathStr.substring(startIdx);
  final endIdx = sub.indexOf(']');
  if (endIdx == -1) return ('', '');

  final content = sub.substring(1, endIdx);
  final parts = content.split(':');
  if (parts.length != 2) return ('', '');

  final startStr = parts[0].trim();
  final endStr = parts[1].trim();

  final int start;
  final int? end;

  if (startStr.isEmpty) {
    start = 0;
  } else {
    start = int.tryParse(startStr) ?? -1;
    if (start == -1) return ('', '');
  }

  if (endStr.isEmpty) {
    end = null;
  } else {
    end = int.tryParse(endStr);
    if (end == null) return ('', '');
  }

  final int? l = end != null ? (end - start < 0 ? 0 : end - start) : null;
  final limitStr = l != null ? '$l' : '';
  final offsetStr = '$start';

  return (limitStr, offsetStr);
}
